From a6404178ac18eb1eb210a06d91749606dcd2b8f4 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Sun, 25 Jun 2023 17:24:49 +0800
Subject: [PATCH] mm/sharepool: Fix softlockup in sp_pte_entry

Call trace:
 dump_backtrace+0x0/0x218
 show_stack+0x2c/0x40
 dump_stack+0x120/0x1a8
 nmi_cpu_backtrace+0x12c/0x138
 nmi_trigger_cpumask_backtrace+0x16c/0x1a8
 arch_trigger_cpumask_backtrace+0x3c/0x78
 rcu_dump_cpu_stacks+0x14c/0x210
 print_cpu_stall+0x1d4/0x3c0
 rcu_sched_clock_irq+0x750/0x9c0
 update_process_times+0x64/0x90
 tick_sched_handle.isra.6+0x34/0x74
 tick_sched_timer+0x50/0xa8
 __hrtimer_run_queues+0x178/0x3f4
 hrtimer_interrupt+0x124/0x430
 arch_timer_handler_virt+0x44/0x70
 handle_percpu_devid_irq+0xb4/0x240
 generic_handle_irq+0x50/0x78
 __handle_domain_irq+0xbc/0x120
 gic_handle_irq+0x120/0x380
 el1_irq+0xbc/0x180
 queued_spin_lock_slowpath+0x54/0x330
 __migration_entry_wait+0xdc/0xe8
 migration_entry_wait+0x64/0x78
 sp_pte_entry+0x15c/0x194
 walk_pgd_range+0x5f4/0x680
 __walk_page_range+0x180/0x1d4
 walk_page_range+0x120/0x198
 __sp_walk_page_range+0x1b0/0x314
 mg_sp_walk_page_range+0xcc/0x290
 dev_sp_walk_page_range+0xb4/0x280 [sharepool_dev]

We have acquired the pagetable lock in the walk_page_range and
migration_entry_wait would acquire it again. Release the lock before
migration_entry_wait to solve the problem.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 78de3d053..e30ba516b 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -2902,13 +2902,17 @@ static int sp_pte_entry(pte_t *pte, unsigned long addr,
 retry:
 	if (unlikely(!pte_present(*pte))) {
 		swp_entry_t entry;
+		spinlock_t *ptl = pte_lockptr(walk->mm, pmd);
 
 		if (pte_none(*pte))
 			goto no_page;
 		entry = pte_to_swp_entry(*pte);
 		if (!is_migration_entry(entry))
 			goto no_page;
+
+		pte_unmap_unlock(pte, ptl);
 		migration_entry_wait(walk->mm, pmd, addr);
+		pte = pte_offset_map_lock(walk->mm, pmd, addr, &ptl);
 		goto retry;
 	}
 
-- 
2.17.1

